scipy/stats/_kde.py

```diff
@@ -596,30 +596,27 @@ def logpdf(self, x):
                 points = reshape(points, (self.d, 1))
                 m = 1
             else:
-                msg = "points have dimension %s, dataset has dimension %s" % (d,
-                    self.d)
+                msg = (f"points have dimension {d}, "
+                       f"dataset has dimension {self.d}")
                 raise ValueError(msg)
 
-        if m >= self.n:
-            # there are more points than data, so loop over data
-            energy = np.empty((self.n, m), dtype=float)
-            for i in range(self.n):
-                diff = self.dataset[:, i, newaxis] - points
-                tdiff = dot(self.inv_cov, diff)
-                energy[i] = sum(diff*tdiff, axis=0)
-            log_to_sum = 2.0 * np.log(self.weights) - self.log_det - energy.T
-            result = logsumexp(0.5 * log_to_sum, axis=1)
+        output_dtype = np.common_type(self.covariance, points)
+        itemsize = np.dtype(output_dtype).itemsize
+        if itemsize == 4:
+            spec = 'float'
+        elif itemsize == 8:
+            spec = 'double'
+        elif itemsize in (12, 16):
+            spec = 'long double'
         else:
-            # loop over points
-            result = np.empty((m,), dtype=float)
-            for i in range(m):
-                diff = self.dataset - points[:, i, newaxis]
-                tdiff = dot(self.inv_cov, diff)
-                energy = sum(diff * tdiff, axis=0)
-                log_to_sum = 2.0 * np.log(self.weights) - self.log_det - energy
-                result[i] = logsumexp(0.5 * log_to_sum)
-
-        return result
+            raise ValueError(
```

Comment:
Can this be tested? Or, if it is already tested, can we include a `match` to make sure the intended message is emitted?

Reply:
Partially, it can be tested. 
I copied it from `evaluate` method so that `logpdf` should be consistent with `evaluate` and I don't have any reason to write this code except for it.

`np.common_type` returns one of `[_nx.half, _nx.single, _nx.double, _nx.longdouble, _nx.csingle, _nx.cdouble, _nx.clongdouble]`. And its `itemsize` is also one of them
```python
import numpy.core.numeric as _nx
print(np.dtype(_nx.half), np.dtype(_nx.half).itemsize)
print(np.dtype(_nx.single), np.dtype(_nx.single).itemsize)
print(np.dtype(_nx.double), np.dtype(_nx.double).itemsize)
print(np.dtype(_nx.longdouble), np.dtype(_nx.longdouble).itemsize)
print(np.dtype(_nx.csingle), np.dtype(_nx.csingle).itemsize)
print(np.dtype(_nx.cdouble), np.dtype(_nx.cdouble).itemsize)
print(np.dtype(_nx.clongdouble), np.dtype(_nx.clongdouble).itemsize)
```
```
float16 2
float32 4
float64 8
float128 16
complex64 8
complex128 16
complex256 32
```
It means if np.dtype is complex256 or float16 it raises an error. 


For complex256 , the test would be like,
```python
def test_pdf_logpdf_dtype_validation():
    xn = (np.random.random(10) + np.random.random(10) * 1j).astype("clongdouble")
    xs = (np.random.random(10) + np.random.random(10) * 1j).astype("clongdouble")
    gkde = stats.gaussian_kde(xn)
    msg = "<class 'numpy.complex256'> has unexpected item size 32"
    with pytest.raises(ValueError, match=msg):
        gkde.logpdf(xs)
```

But, for float16 it can't be tested. `gkde.covariance` is float64 so np.common_type returns float64 even if an array's dtype is float16.